warcprox/writer.py

```diff
@@ -22,178 +22,220 @@
-                        self._f_finalname)
-                self.logger.debug(
-                        'warcinfo_record.headers=%s', warcinfo_record.headers)
-                warcinfo_record.write_to(self._f, gzip=self.gzip)
+    def open(self, serial):
+        if not os.path.exists(self.directory):
+            self.logger.info(
+                    "warc destination directory %s doesn't exist, creating it",
+                    self.directory)
+            os.mkdir(self.directory)
+
+        self.finalname = self.next_filename(serial)
+        self.path = os.path.sep.join(
+                [self.directory, self.finalname + self.open_suffix])
+
+        self.f = open(self.path, 'wb')
+        # if no '.open' suffix is used for WARC, acquire an exclusive
+        # file lock.
+        if self.open_suffix == '':
+            try:
+                fcntl.lockf(self.f, fcntl.LOCK_EX | fcntl.LOCK_NB)
+            except IOError as exc:
+                self.logger.error(
+                        'could not lock file %s (%s)', self.path, exc)
+        return self.f
+
+    def close(self):
+        if self.path:
+            self.logger.trace('closing %s', self.finalname)
+            if self.open_suffix == '':
+                try:
+                    fcntl.lockf(self.f, fcntl.LOCK_UN)
+                except IOError as exc:
+                    self.logger.error(
+                            'could not unlock file %s (%s)', self.path, exc)
+            self.f.close()
+            finalpath = os.path.sep.join(
+                    [self.directory, self.finalname])
+            os.rename(self.path, finalpath)
+
+            self.path = None
+            self.f = None
+
+    def maybe_idle_rollover(self):
+        if (self.path and self.rollover_idle_time
+                and self.rollover_idle_time > 0
+                and time.time() - self.last_activity > self.rollover_idle_time):
+            self.logger.info(
+                    'rolling over %s after %0.1f seconds idle',
+                    self.finalname, time.time() - self.last_activity)
+            self.close()
+
+    def maybe_size_rollover(self):
+        if self.path and os.path.getsize(self.path) > self.rollover_size:
+            self.logger.info(
+                    'rolling over %s because it has reached %s bytes in size',
+                    self.finalname, os.path.getsize(self.path))
+            self.close()
+
+class WarcWriter:
+    logger = logging.getLogger('warcprox.writer.WarcWriter')
 
+    def __init__(self, options=warcprox.Options()):
+        self.options = options
+
+        self.gzip = options.gzip or False
+        self.record_builder = warcprox.warc.WarcRecordBuilder(
+                digest_algorithm=options.digest_algorithm or 'sha1',
+                base32=options.base32)
+
+        self._available_warcs = queue.Queue()
+        self._warc_count = 0
+        self._warc_count_lock = threading.Lock()
```

Comment:
I just note that we previously used `threading.RLock`, is there a reason for changing or maybe I caught something here.

Reply:
Don't think we need this lock to be reentrant (the old one may have needed reentrancy)